lib/maintenance/MaintenanceManager.js

```diff
@@ -0,0 +1,102 @@
+const _ = require('lodash');
+const Promise = require('bluebird');
+const utils = require('../utils');
+const Repository = require('../db').Repository;
+const logger = require('../logger');
+const CONST = require('../constants');
+const errors = require('../errors');
+const config = require('../config');
+
+class MaintenanceManager {
+
+  startMaintenace(maintenanceInfo, user) {
+    return Promise.try(() => {
+      if (maintenanceInfo.progress === undefined) {
+        maintenanceInfo.progress = [`Service-fabrik maintenace mode is being initiated ${new Date()}`];
+      } else {
+        maintenanceInfo.progress = [`${maintenanceInfo.progress} at ${new Date()}`];
+      }
+      if (maintenanceInfo.fromVersion === undefined) {
+        maintenanceInfo.fromVersion = `NA_${new Date().getTime()}`;
+      }
+      if (maintenanceInfo.toVersion === undefined) {
+        maintenanceInfo.toVersion = `NA_${new Date().getTime()}`;
+      }
+      maintenanceInfo.state = CONST.OPERATION.IN_PROGRESS;
+      maintenanceInfo.completedAt = null;
+      logger.info('Going to save maintenance detail:', maintenanceInfo);
+      return Repository.save(CONST.DB_MODEL.MAINTENANCE_DETAIL, maintenanceInfo, user);
+    });
+  }
+
+  updateMaintenace(progressInfo, state, user) {
+    const criteria = {
+      state: `${CONST.OPERATION.IN_PROGRESS}`
+    };
+    logger.info('searching by criteria: ', criteria);
+    return Repository
+      .findOne(CONST.DB_MODEL.MAINTENANCE_DETAIL, criteria)
+      .then((model) => {
+        if (_.isEmpty(model)) {
+          throw new errors.BadRequest(`System not in maitenance mode`);
+        }
+        logger.info('found system maintenance :', model);
+        if (progressInfo === undefined || _.isEmpty(progressInfo.trim())) {
+          throw new errors.BadRequest(`Progressinfo is mandatory`);
+        } else {
+          model.progress.push(`${progressInfo} at ${new Date()}`);
+        }
+        if (state !== undefined) {
+          if (_.chain(CONST.OPERATION).valuesIn().indexOf(state).value() === -1) {
+            throw new errors.BadRequest(`Maintenance state can be only one of these values : ${_.valuesIn(CONST.OPERATION)}`);
+          }
+          if (utils.isServiceFabrikOperationFinished(state)) {
+            model.completedAt = new Date();
+          }
+          model.state = state;
+        }
+        return Repository.save(CONST.DB_MODEL.MAINTENANCE_DETAIL, model, user);
```

Comment:
You can put a logger before this saying updating with state, so that it completes the circle of start and update of maintenance.

Reply:
completing the circle :-)